### PR TITLE
[Form] allow form types + form type extensions + form type guessers to be private services

### PR DESCRIPTION
**Pull request by hhamon**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | ~

This pull request is about making internal form services (aka form types, form type extensions and form type guessers) private. They used to be public until Symfony 3.2 for one valid reason: lazyness. However, Symfony 3.3 now comes with built-in mechanism to support effective lazy loading of private services with service locators and proxies.

This PR makes the `DependencyInjectionExtension` class of the `Form` component leverage these new DI component mechanisms. Form types, form type extensions and form type guessers can now be declared private as a best practice. We decided to make these services private as of Symfony 3.3 and of course it would break BC. But this PR introduces a BC layer using a Symfony trick to keep internal form services public. The service container currently has a known issue where private services are not really private if they're referenced by at least two other services in the container. We use this trick to maintain the legacy services public even though the new API relies on private ones. This trick is done thanks to the `deprecated.form.registry` and `deprecated.form.registry.csrf` fake services that will be removed in Symfony 4.0.

### CODE REVIEW

**Comment by hhamon** [comment]
Related to #21553.

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 1]
```diff
@@ -78,18 +85,20 @@ public function process(ContainerBuilder $container)
...
         foreach ($guessers as $serviceId) {
-            $serviceDefinition = $container->getDefinition($serviceId);
-            if (!$serviceDefinition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form type guessers are lazy-loaded.', $serviceId));
-            }
+            // Add form type guesser service to the service locator
+            $servicesMap[$serviceId] = new Reference($serviceId);
         }
 
+        $definition->replaceArgument(0, new ServiceLocatorArgument($servicesMap));
```
Comment (chalasr) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: I'm wondering if mixing types/guessers/extensions in the same locator is a good thing here, as we have still 3 constructor arguments that's a bit confusing to me.
Instead, couldn't `guesserServices` be an `IteratorArgument` of guessers? It seems we only iterate over them to build a chain one, no need for having them identified right?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 2]
```diff
@@ -11,55 +11,87 @@
...
+    private $typeContainer;
+    private $typeExtensionServices;
+    private $guesserServices;
+
+    /**
+     * Constructor.
+     *
+     * @param ContainerInterface $typeContainer
+     * @param iterable[]         $typeExtensionServices
+     * @param iterable[]         $guesserServices
```
Comment (chalasr) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: `iterable[]` make me understand that the value should be an array of iterables.
`iterable|FormTypeGuesser[]` instead?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 3]
```diff
@@ -82,8 +114,13 @@ public function getTypeGuesser()
...
 
-            foreach ($this->guesserServiceIds as $serviceId) {
-                $guessers[] = $this->container->get($serviceId);
+            foreach ($this->guesserServices as $serviceId => $service) {
+                if (is_string($service)) {
+                    $serviceId = $service;
+                    $service = $this->typeContainer->get($serviceId);
+                }
+
+                $guessers[] = $service;
```
Comment (chalasr) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: If fact, guessers do not need to be instantiated at this point. 
I think we should register `FormTypeGuesserChain` as a service, pass it guessers as an IteratorArgument and remove both `$guesserServices` and `$guesserServiceIds` in favor of a `FormTypeGuesserInterface $guesser` here. Not sure how it would be hard to keep it BC though

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php (hunk 1 above)
Comment (hhamon) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Yes It's on the way! This is what I'm currently doing :)

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 2 above)
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Yes it can be an array or an iterator.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 3 above)
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Yes that's probably a good thing! I'll check this tomorrow with @nicolas-grekas.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml [hunk 4]
```diff
@@ -30,7 +30,7 @@
 
         <!-- DependencyInjectionExtension -->
         <service id="form.extension" class="Symfony\Component\Form\Extension\DependencyInjection\DependencyInjectionExtension" public="false">
-            <argument type="service" id="service_container" />
+            <argument type="service-locator" />
```
Comment (hhamon) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: Yep good catch!

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php (hunk 1 above)
Comment (ogizanagi) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Why not 2 service locators, and one iterator?

So the `DependencyInjectionExtension::__construct()` method signature would be:

```php
public function __construct(ContainerInterface $typesLocator, ContainerInterface $typeExtensionsLocator, iterator $guessers)
```

Might not be the easiest for BC though.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 3 above)
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Not sure we need `FormTypeGuesserChain` to be a service, but simply pass it the guessers iterator (so allow `FormTypeGuesserChain` to accept iterable).
IMHO the extension is in charge of building the whole thing and the use of the `FormTypeGuesserChain` is an implementation detail. There is no need to pass the guesser chain as a dependency.

There is an issue with `FormTypeGuesserChain`, though... its constructor. It iterates over the guessers and this would trigger the loading too soon anyway.

I'm not sure lazy loading here is a big deal though, as `FormExtensionInterface::getTypeGuesser()` is called in `FormRegistryInterface::getTypeGuesser()`, called in `Formfactory::createBuilderForProperty()` and is used directly to guess types by iterating over every guessers.

File: src/Symfony/Component/Form/Tests/DependencyInjection/FormPassTest.php [hunk 5]
```diff
@@ -35,47 +34,33 @@ public function testDoNothingIfFormExtensionNotLoaded()
...
 
         $container->compile();
 
         $extDefinition = $container->getDefinition('form.extension');
 
-        $this->assertEquals(array(
-            __CLASS__.'_Type1' => 'my.type1',
-            __CLASS__.'_Type2' => 'my.type2',
-        ), $extDefinition->getArgument(1));
+        $this->assertSame(
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Tests/DependencyInjection/FormPassTest.php]: Should be reverted too I think.

File: src/Symfony/Component/Form/Tests/DependencyInjection/FormPassTest.php [hunk 6]
```diff
@@ -146,67 +123,70 @@ public function testAddTaggedFormTypeExtensionWithoutExtendedTypeAttribute()
...
         $container->compile();
 
         $extDefinition = $container->getDefinition('form.extension');
 
-        $this->assertSame(array(
-            'my.guesser1',
-            'my.guesser2',
-        ), $extDefinition->getArgument(3));
+        $this->assertSame(
+            array(
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Tests/DependencyInjection/FormPassTest.php]: same

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php (hunk 1 above)
Comment (ogizanagi) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Oh, actually `$typeExtensionServiceIds` cannot easily be a locator, as it is a map of multiple services by type and symfony's `ServiceLocator` does not handle collection of services as values.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 3 above)
Comment (chalasr) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: > IMHO the extension is in charge of building the whole thing

Too many responsibilities to me.

> and the use of the FormTypeGuesserChain is an implementation detail. There is no need to pass the guesser chain as a dependency.

The fact the guessers are multiple services is a detail IMO. At the end we have only one `FormTypeGuesserInterface $guesser`, the class could be usable with an iterable or a simple guesser, which looks better to me.

> There is an issue with FormTypeGuesserChain, though... its constructor. It iterates over the guessers and this would trigger the loading too soon anyway.

Just validation/merge, can be moved at compile time I think.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/FormPassTest.php [hunk 7]
```diff
@@ -37,47 +36,33 @@ public function testDoNothingIfFormExtensionNotLoaded()
...
 
         $container->compile();
 
         $extDefinition = $container->getDefinition('form.extension');
 
-        $this->assertEquals(array(
-            __CLASS__.'_Type1' => 'my.type1',
-            __CLASS__.'_Type2' => 'my.type2',
-        ), $extDefinition->getArgument(1));
+        $this->assertSame(
```
Comment (hhamon) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/FormPassTest.php]: Why?

Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/FormPassTest.php]: The code style and the `assertSame` change isn't related to this PR, right?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 3 above)
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: I thought about deprecating the whole class and replacing it with a new one. But after discussing with @nicolas-grekas it requires too much efforts then to maintain both implementations. This is why we choose to adapt the class itself. One class to maintain and one tests suite. It's just a bit more difficult to read and understand the code at the moment because of the mixed two layers.

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/FormPassTest.php [hunk 8]
```diff
@@ -37,47 +38,37 @@ public function testDoNothingIfFormExtensionNotLoaded()
...
 
         $extDefinition = $container->getDefinition('form.extension');
-
-        $this->assertEquals(array(
-            __CLASS__.'_Type1' => 'my.type1',
-            __CLASS__.'_Type2' => 'my.type2',
-        ), $extDefinition->getArgument(1));
+        $locator = $extDefinition->getArgument(0);
+
+        $this->assertInstanceOf(ServiceLocatorArgument::class, $locator);
```
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/FormPassTest.php]: No need for `assertInstanceOf` as `assertEquals` fails on an type-class mismatch.

Comment (hhamon) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/FormPassTest.php]: Fixed. thanks.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 3 above)
Comment (chalasr) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: > It's just a bit more difficult to read and understand the code at the moment because of the mixed two layers.

Indeed :)  
I'm unable to say if the BC layer is correct right now, but the approach looks better!

Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Unit tests prove the BC layer is guaranteed ;)

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 9]
```diff
@@ -47,28 +50,25 @@ public function process(ContainerBuilder $container)
...
         foreach ($container->findTaggedServiceIds($this->formTypeTag) as $serviceId => $tag) {
             $serviceDefinition = $container->getDefinition($serviceId);
-            if (!$serviceDefinition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form types are lazy-loaded.', $serviceId));
-            }
-
-            // Support type access by FQCN
-            $types[$serviceDefinition->getClass()] = $serviceId;
+            // Add form type service to the service locator
+            $servicesMap[$serviceId] = new Reference($serviceId);
```
Comment (hhamon) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Ok thanks!

Comment (hhamon) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Fixed.

**Comment by hhamon** [comment]
Woot woot! Tests are back to green 🎆

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 10]
```diff
@@ -46,50 +50,69 @@ public function process(ContainerBuilder $container)
...
-        $definition->replaceArgument(1, $types);
+        return new ServiceLocatorArgument($servicesMap);
+    }
 
+    private function processFormTypeExtensions(ContainerBuilder $container)
+    {
         $typeExtensions = array();
-
-        foreach ($this->findAndSortTaggedServices($this->formTypeExtensionTag, $container) as $reference) {
+        foreach ($this->findAndSortTaggedServices($this->formTypeExtensionTag,
```
Comment (sstok) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Wrong CS 😜

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 11]
```diff
@@ -46,50 +50,69 @@ public function process(ContainerBuilder $container)
...
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form type extensions are lazy-loaded.', $serviceId));
-            }
 
             $tag = $serviceDefinition->getTag($this->formTypeExtensionTag);
             if (isset($tag[0]['extended_type'])) {
                 $extendedType = $tag[0]['extended_type'];
             } else {
-                throw new InvalidArgumentException(sprintf('"%s" tagged services must have the extended type configured using the extended_type/extended-type attribute, none was configured for the "%s" service.', $this->formTypeExtensionTag, $serviceId));
+                throw new InvalidArgumentException(sprintf('"%s" tagged services must have the extended type configured using the extended_type/extended-type attribute, none was configured for the "%s" service.',
+                    $this->formTypeExtensionTag, $serviceId));
```
Comment (yceruto) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: should be one line as before

Comment (hhamon) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Oups thanks!

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php (hunk 10 above)
Comment (hhamon) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Thanks!

**Comment by hhamon** [comment]
ping @symfony/deciders for review please.

**Comment by hhamon** [comment]
Review from Symfony Core Team member is welcome.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 12]
```diff
@@ -11,55 +11,87 @@
...
+     * @param array|\Traversable $guesserServices
+     */
+    public function __construct(ContainerInterface $typeContainer, array $typeExtensionServices, $guesserServices, array $guesserServiceIds = null)
     {
-        $this->container = $container;
-        $this->typeServiceIds = $typeServiceIds;
-        $this->typeExtensionServiceIds = $typeExtensionServiceIds;
-        $this->guesserServiceIds = $guesserServiceIds;
+        if (null !== $guesserServiceIds) {
+            @trigger_error(sprintf('Passing four arguments to the %s::__construct() method is deprecated since Symfony 3.3 and will be disallowed in Symfony 4.0. The new prototype only accepts three arguments.', __CLASS__), E_USER_DEPRECATED);
```
Comment (javiereguiluz) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: `The new prototype` -> `The new constructor` ?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 13]
```diff
@@ -11,55 +11,87 @@
...
+            $this->guesserServiceIds = $guesserServiceIds;
+            $this->typeServiceIds = $typeExtensionServices;
+        } else {
+            $this->typeContainer = $typeContainer;
```
Comment (javiereguiluz) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Can we remove this line because it's already executed 3 lines below?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 14]
```diff
@@ -11,55 +11,87 @@
...
+        $this->typeExtensionServices = $typeExtensionServices;
+        $this->guesserServices = $guesserServices;
     }
 
     public function getType($name)
     {
-        if (!isset($this->typeServiceIds[$name])) {
+        if (null !== $this->guesserServiceIds) {
+            if (!isset($this->typeServiceIds[$name])) {
+                throw new InvalidArgumentException(sprintf('The field type "%s" is not registered with the service container.', $name));
```
Comment (javiereguiluz) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: `registered with` -> `registered in` ?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 13 above)
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Good catch! Thanks!

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 12 above)
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Fixed.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 14 above)
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Fixed.

**Comment by hhamon** [comment]
@javiereguiluz all fixed. thanks.

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 15]
```diff
@@ -80,16 +91,26 @@ public function process(ContainerBuilder $container)
             $typeExtensions[$extendedType][] = $serviceId;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: shouldn't we just do `$typeExtensions[$extendedType][] = new Reference($serviceId);` here and remove the `array_map` below?

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 16]
```diff
@@ -80,16 +91,26 @@ public function process(ContainerBuilder $container)
...
-            if (!$serviceDefinition->isPublic()) {
-                throw new InvalidArgumentException(sprintf('The service "%s" must be public as form type guessers are lazy-loaded.', $serviceId));
-            }
+        return $allExtensions;
+    }
+
+    private function processFormTypeGuessers(ContainerBuilder $container)
+    {
+        $guessers = array();
+        foreach (array_keys($container->findTaggedServiceIds($this->formTypeGuesserTag)) as $serviceId) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: no need for array_keys, it only creates an intermediary array for no reason, let's do 
`foreach ($container->findTaggedServiceIds($this->formTypeGuesserTag) as $serviceId => $tags) {`

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 17]
```diff
@@ -11,55 +11,85 @@
...
+    private $typeContainer;
+    private $typeExtensionServices;
+    private $guesserServices;
+
+    /**
+     * Constructor.
+     *
+     * @param ContainerInterface $typeContainer
+     * @param iterable[]         $typeExtensionServices
+     * @param array|\Traversable $guesserServices
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: `array|\Traversable` => iterable?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 18]
```diff
@@ -11,55 +11,85 @@
...
     public function getTypeExtensions($name)
     {
         $extensions = array();
 
-        if (isset($this->typeExtensionServiceIds[$name])) {
-            foreach ($this->typeExtensionServiceIds[$name] as $serviceId) {
-                $extensions[] = $extension = $this->container->get($serviceId);
+        if (isset($this->typeExtensionServices[$name])) {
+            foreach ($this->typeExtensionServices[$name] as $extensionId => $extension) {
+                if (is_string($extension)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: add a note saying this "if" should be removed in 4.0 (shouldn't it?)

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 19]
```diff
@@ -82,8 +112,13 @@ public function getTypeGuesser()
             $this->guesserLoaded = true;
             $guessers = array();
 
-            foreach ($this->guesserServiceIds as $serviceId) {
-                $guessers[] = $this->container->get($serviceId);
+            foreach ($this->guesserServices as $serviceId => $service) {
+                if (is_string($service)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: removal notice for 4.0 also?

File: src/Symfony/Component/Form/Tests/DependencyInjection/FormPassTest.php [hunk 20]
```diff
@@ -12,10 +12,12 @@
 namespace Symfony\Component\Form\Tests\DependencyInjection;
 
 use PHPUnit\Framework\TestCase;
+use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
+use Symfony\Component\DependencyInjection\Argument\ServiceLocatorArgument;
+use Symfony\Component\DependencyInjection\Reference;
 use Symfony\Component\Form\DependencyInjection\FormPass;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Definition;
-use Symfony\Component\DependencyInjection\Reference;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/DependencyInjection/FormPassTest.php]: Keep Reference here, no reordering

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php [hunk 21]
```diff
@@ -46,29 +50,36 @@ public function process(ContainerBuilder $container)
         }
 
         $definition = $container->getDefinition($this->formExtensionService);
```
Comment (HeahDude) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: I just realised that I never posted my comment from January 26th on https://github.com/symfony/symfony/pull/21283 by forgetting to publish the review :(. But this is broken, we need to ensure that the defined extension is an instance of `DependencyInjectionExtension`.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 22]
```diff
@@ -11,55 +11,85 @@
...
-    public function __construct(ContainerInterface $container, array $typeServiceIds, array $typeExtensionServiceIds, array $guesserServiceIds)
+    private $typeContainer;
+    private $typeExtensionServices;
+    private $guesserServices;
+
+    /**
+     * Constructor.
+     *
+     * @param ContainerInterface $typeContainer
+     * @param iterable[]         $typeExtensionServices
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: `array`?

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 23]
```diff
@@ -11,55 +11,85 @@
...
+        $this->typeContainer = $typeContainer;
+        $this->typeExtensionServices = $typeExtensionServices;
+        $this->guesserServices = $guesserServices;
     }
 
     public function getType($name)
     {
-        if (!isset($this->typeServiceIds[$name])) {
-            throw new InvalidArgumentException(sprintf('The field type "%s" is not registered with the service container.', $name));
+        if (null !== $this->guesserServiceIds) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: I would add a comment `// BC` to ease readability of the layer, also to help remembering to remove it in 4.0.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php [hunk 24]
```diff
@@ -11,55 +11,85 @@
...
         }
 
-        return $this->container->get($this->typeServiceIds[$name]);
+        return $this->typeContainer->get($name);
     }
 
     public function hasType($name)
     {
-        return isset($this->typeServiceIds[$name]);
+        if (null !== $this->guesserServiceIds) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: Same here.

**Review by HeahDude — COMMENTED** [review]
Thank you for working on this @hhamon :)

**Comment by hhamon** [comment]
@nicolas-grekas @HeahDude comments were addressed. I'll squash after your new view.

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php (hunk 21 above)
Comment (hhamon) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: This should be addressed in a separate PR.

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 22 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: iterable[] is more precise: array of iterable

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 23 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: notice moved near the properties declaration

File: src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php (hunk 24 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php]: notice moved near the properties declaration

**Review by HeahDude — APPROVED** [review]
👍

File: src/Symfony/Component/Form/composer.json [hunk 25]
```diff
@@ -25,8 +25,9 @@
     },
     "require-dev": {
         "doctrine/collections": "~1.0",
+        "psr/container": "^1.0",
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/composer.json]: not required (as symfony/dependency-injection already requires it)

**Review by nicolas-grekas — APPROVED** [review]
:+1:

File: src/Symfony/Component/Form/DependencyInjection/FormPass.php (hunk 21 above)
Comment (chalasr) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: @HeahDude what do you recommend in case it is not? Throwing or just returning? I'll fix it

Comment (HeahDude) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: IMHO `form.extension` should be a reserved service, required for the `FormPass` to do the job, so an exception should be thrown in the case it is defined but not an instance of `DependencyInjectionExtension`.

What do you think __@symfony/deciders__?

Comment (HeahDude) [inline src/Symfony/Component/Form/DependencyInjection/FormPass.php]: Ok, thank you for the demonstration ;), it indeed looks like this extra check is not needed.

**Comment by hhamon** [comment]
@stof @webmozart can I have your review please?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml [hunk 26]
```diff
@@ -30,17 +30,13 @@
...
-            <!-- All services with tag "form.type_guesser" are inserted here by FormPass -->
-            <argument type="collection" />
+            <argument type="service-locator" /><!-- All services with tag "form.type" are stored in a service locator by FormPass -->
+            <argument type="collection" /><!-- All services with tag "form.type_extension" are stored here by FormPass -->
+            <argument type="iterator" /><!-- All services with tag "form.type_guesser" are stored here by FormPass -->
         </service>
 
         <!-- ValidatorTypeGuesser -->
-        <service id="form.type_guesser.validator" class="Symfony\Component\Form\Extension\Validator\ValidatorTypeGuesser">
+        <service id="form.type_guesser.validator" class="Symfony\Component\Form\Extension\Validator\ValidatorTypeGuesser" public="false">
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: see PR description for why this is not a BC break (same for all added `public="false"`)

**Review by chalasr — APPROVED** [review]
👍  (needs rebase)

**Comment by fabpot** [comment]
Thank you @hhamon.
